File: javascript/1_Arrays&Hashing/04_GroupAnagrams.js
strs = ["eat","tea","tan","ate","nat","bat"]

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
const groupAnagrams = (strs) => {
    if (!strs.length) return []
    const map = new Map()

    for (const s of strs) {
        const frequency = new Array(26).fill(0);

        for (const char of s) {
            const charCode = char.charCodeAt(0) - 'a'.charCodeAt(0);
            frequency[charCode]++;
        }

        const hash = frequency.toString();
        const values = map.get(hash) || [];

        values.push(s);
        map.set(hash, values);
    }

    return [ ...map.values() ];
};

console.log(groupAnagrams(strs))

File: python/1_Arrays&Hashing/04_GroupAnagrams.py
from collections import defaultdict
from typing import List


class Solution:
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]

    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = defaultdict(list)

        print(res)

        for s in strs:
            count = [0] * 26  # binary array of 26 length

            for c in s:
                count[ord(c) - ord("a")] += 1

            res[tuple(count)].append(s)





        return res.values()


solution = Solution()
print(solution.groupAnagrams(solution.strs))
